packages/client/lib/sync/execution/vmexecution.ts

```diff
@@ -126,14 +147,26 @@ export class VMExecution extends Execution {
                 `Set back hardfork along block deletion number=${blockNumber} hash=${hash} old=${this.hardfork} new=${hardfork}`
               )
               this.config.execCommon.setHardforkByBlockNumber(blockNumber)
-            }
-            throw error
+            }*/
+            // Option a): set iterator head to the parent block so that an
+            // error can repeatedly processed for debugging
+            const blockNumber = block.header.number.toNumber()
+            const hash = short(block.hash())
+            this.config.logger.warn(`Execution of block number=${blockNumber} hash=${hash} failed`)
+            this.emit('error', error)
+            errorBlock = block
           }
         },
         this.NUM_BLOCKS_PER_ITERATION
       )
       numExecuted = (await this.vmPromise) as number
 
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+      if (errorBlock) {
+        await this.chain.blockchain.setIteratorHead('vm', (errorBlock as Block).header.parentHash)
```

Comment:
We are not deleting the block? If you would restart sync now, then it will re-execute the wrong block, right?

Reply:
Yes, that's correct.

The main point here is to bring us in a better debugging position. As stated in the comment above, **IF** there is a consensus error like in #1076 we just have to fix and do a new release, there is no way to develop around this case.

I only implemented the case from above in this PR since for some time I very much assume this will be our main (if not: only) concern from both cases. If there is an "unintentionally bad peer" which is just serving malformed blocks by software error this is now covered by having activated the block validation on blockchain (not sure though if the sync will recover properly on such a case, but this is then something for another PR).

On `mainnet` it is highly unlikely that someone will try to trick us on a deviating chain path in the lower block number range being so far away from the latest mainnet block (so, let's say: someone is sending out a malformed (but validated) block nr. 1.000.000 just to confuse our few 1,2,3 client instances. On the PoA chains (so I would say: these are the two cases we both target primarily atm) bad blocks from attackers are just no thing at all. So we can safely concentrate on the consensus part right now, on mainnet being able to sync close to the chain head is still at least weeks - or rather months - away I would assume (if we will get there at all, many unknowns on the performance side).